### PR TITLE
KT-17110 Rewrite branches targeting other branches

### PR DESCRIPTION
**Pull request by mikaelpeltier**
- Rewrite branches targeting other branches to target directly the
final destination to avoid runtime performance penalties.

Fix of https://youtrack.jetbrains.com/issue/KT-17110

### CODE REVIEW

**Comment by mikaelpeltier** [comment]
This PR was tested by:
./gradlew dist
./gradlew jvmCompilerTest

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt [hunk 1]
```diff
@@ -22,31 +22,76 @@ import org.jetbrains.org.objectweb.asm.Opcodes
...
+     *      goto Label2 (must not be removed due to the previous instruction that can fallthrough on this goto)
+     *      ...
      */
     override fun transform(internalClassName: String, methodNode: MethodNode) {
         val insns = methodNode.instructions.toArray().apply { reverse() }
-        val insnsToRemove = arrayListOf<AbstractInsnNode>()
-
+        var insnsToRemove = arrayListOf<AbstractInsnNode>()
         val currentLabels = hashSetOf<LabelNode>()
+        var labelsToReplace = hashMapOf<LabelNode, JumpInsnNode>()
```
Comment (dnpetrov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: `labelsToReplace` can be a `val`.

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt [hunk 2]
```diff
@@ -22,31 +22,76 @@ import org.jetbrains.org.objectweb.asm.Opcodes
...
+     *      ...
+     *      Label:
+     *      goto Label2 (must not be removed due to the previous instruction that can fallthrough on this goto)
+     *      ...
      */
     override fun transform(internalClassName: String, methodNode: MethodNode) {
         val insns = methodNode.instructions.toArray().apply { reverse() }
-        val insnsToRemove = arrayListOf<AbstractInsnNode>()
-
+        var insnsToRemove = arrayListOf<AbstractInsnNode>()
```
Comment (dnpetrov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: `insnsToRemove` can be a `val`.

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt [hunk 3]
```diff
@@ -22,31 +22,76 @@ import org.jetbrains.org.objectweb.asm.Opcodes
...
+                insn.isMeaningful -> {
                     currentLabels.clear()
+                    pendingGoto = null
+                }
             }
         }
 
+        // Rewrite branch instructions.
+        if (!labelsToReplace.isEmpty()) {
+            insns.filter { it is JumpInsnNode }.forEach { rewriteLabelIfNeeded(it as JumpInsnNode, labelsToReplace) }
```
Comment (dnpetrov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Use `filterIsInstance`.

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt [hunk 4]
```diff
@@ -22,31 +22,76 @@ import org.jetbrains.org.objectweb.asm.Opcodes
...
         for (insnToRemove in insnsToRemove) {
             methodNode.instructions.remove(insnToRemove)
         }
     }
+
+    private fun rewriteLabelIfNeeded(
+        jumpInsn: JumpInsnNode,
+        labelsToReplace: HashMap<LabelNode, JumpInsnNode>
+    ) {
+        getLastTargetJumpInsn(jumpInsn, labelsToReplace).takeIf { it.label != jumpInsn.label }?.let {
```
Comment (dnpetrov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Minor: I'd rather just use `if`. I find `takeIf` + `?.let` a bit more convoluted way to say basically the same thing: do something with a value if condition is met.

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt (hunk 1 above)
Comment (mikaelpeltier) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Done

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt (hunk 2 above)
Comment (mikaelpeltier) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Done

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt (hunk 3 above)
Comment (mikaelpeltier) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Done

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt (hunk 4 above)
Comment (mikaelpeltier) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt]: Done

**Comment by mikaelpeltier** [comment]
Pushed to master by dnpetrov.

**Pull request closed without merging**